---
title: "Ectoine gene BLASTing"
output: html_notebook
---

```{r setup, include=FALSE}
#Set working directory to project root
knitr::opts_knit$set(root.dir = here::here())

#Load packages
library(tidyverse)
```

## Profile Ectoine Genes

### Contig taxonomy from kraken
```{r}
contig_tax <- read_tsv("data/kraken2_contigs/gtdb/translated_out.txt",col_names = c("contig","contig_taxonomy"))
```

### Blast results for Ectoine genes, summarized with contig taxonomy from Kraken
```{r}
gene_data <- read_csv("data/reference/ectoine_metabolism_genes.csv")

blast6_cols <- c("Query", "Subject", "Percent_ID", "Aln_Len", "Mismatch", "Gap", "Query_Start", "Query_Stop", "Subject_Start", "Subject_Stop", "E-Value", "Bit_Score", "percent_of_query_aligned", "subject_title","aligned_qeuery_seq", "aligned_subject_seq")

contig_coverage <- read_tsv("data/contig_coverage.tsv") %>% 
  rename_all(~str_remove(.x,".* "))

blast_res <- read_tsv("data/BLAST/ectoine_genes_BLAST.blastn", skip = 1,col_types = "ccdiddiiiidddccc", col_names = blast6_cols) %>% 
  left_join(gene_data) %>% 
  left_join(contig_tax %>% rename(subject_title = "contig")) %>% 
  left_join(contig_coverage %>% rename(subject_title = "Contig"))

# Filter using EggNog Mapper filtering criteria
blast_res_filtered <- blast_res %>% 
  filter(`E-Value` < 0.001, Bit_Score > 60, Percent_ID > 40, percent_of_query_aligned > 20) 

tax_levels <- c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")

blast_res_filtered %>% 
  mutate(contig_taxonomy = str_remove(contig_taxonomy,"\\|g__.*"),
         contig_taxonomy = str_remove_all(contig_taxonomy,"\\|[a-z]__unassigned"),
         contig_taxonomy = str_replace_all(contig_taxonomy,"\\|o","\no"),
         contig_taxonomy = str_replace_all(contig_taxonomy,"\\|"," | "),
         contig_taxonomy = if_else(is.na(contig_taxonomy),"Unclassified",contig_taxonomy)) %>% 
  group_by(contig_taxonomy,geneID,synth_or_deg) %>% 
  summarise(RPKM = mean(RPKM),
            Mean = mean(Mean)) %>%
  ggplot(aes(geneID, contig_taxonomy, fill = Mean)) +
  geom_tile() + 
  scale_fill_viridis_c() +
  facet_grid(~synth_or_deg,scales = "free_x",space = "free_x") +
  labs(y = NULL, x= "Gene" , fill ="Mean\nCoverage") +
  theme_bw()

ggsave("results/Ectoine_gene_coverage.png",width = 3, height = 3, dpi = 600, scale = 3)


blast_res_filtered %>% 
  mutate(contig_taxonomy = str_remove(contig_taxonomy,"\\|o__.*"),
         contig_taxonomy = str_remove_all(contig_taxonomy,"\\|[a-z]__unassigned"),
         contig_taxonomy = str_replace_all(contig_taxonomy,"\\|o","\no"),
         contig_taxonomy = str_replace_all(contig_taxonomy,"\\|"," | "),
         contig_taxonomy = if_else(is.na(contig_taxonomy),"Unclassified",contig_taxonomy)) %>% 
  group_by(contig_taxonomy,geneID,synth_or_deg) %>% 
  summarise(RPKM = mean(RPKM),
            Mean = mean(Mean)) %>% ungroup() %>% group_by(contig_taxonomy) %>% 
  mutate(contig_taxonomy = if_else(contig_taxonomy == "k__Bacteria","Other Bacteria",contig_taxonomy),
         contig_taxonomy = str_remove(contig_taxonomy, "k__"),
         num_genes = sum(Mean > 0)) %>% 
  #ggplot(aes(geneID, factor(contig_taxonomy,levels = tax_order, ordered = TRUE) , fill = Mean)) +
  ggplot(aes(geneID, contig_taxonomy, fill = Mean)) +
  geom_tile() + 
  scale_fill_viridis_c() +
  facet_grid(~synth_or_deg,scales = "free_x",space = "free_x") +
  labs(y = NULL, x= "Gene" , fill ="Mean Contig\nRead Coverage") +
  theme_bw() +
  scale_y_discrete(limits=rev) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())

ggsave("results/Ectoine_gene_coverage_to_Class.png",width = 3, height = 1.5, scale = 3.5, type = "cairo")
ggsave("results/Ectoine_gene_coverage_to_Class.svg",width = 3, height = 1.5, dpi = 600, scale = 3.5)


blast_res_filtered %>% 
  filter(Query_tax =="Halobacillus halophilus") %>% 
  mutate(contig_taxonomy = str_remove(contig_taxonomy,"\\|o__.*"),
         contig_taxonomy = str_remove_all(contig_taxonomy,"\\|[a-z]__unassigned"),
         contig_taxonomy = str_replace_all(contig_taxonomy,"\\|o","\no"),
         contig_taxonomy = str_replace_all(contig_taxonomy,"\\|"," | "),
         contig_taxonomy = if_else(is.na(contig_taxonomy),"Unclassified",contig_taxonomy)) %>% 
  group_by(contig_taxonomy,geneID,synth_or_deg) %>% 
  summarise(RPKM = mean(RPKM),
            Mean = mean(Mean)) %>% ungroup() %>% group_by(contig_taxonomy) %>% 
  mutate(contig_taxonomy = if_else(contig_taxonomy == "k__Bacteria","Other Bacteria",contig_taxonomy),
         contig_taxonomy = str_remove(contig_taxonomy, "k__"),
         num_genes = sum(Mean > 0)) %>% 
  ggplot(aes(geneID, reorder(contig_taxonomy,num_genes) , fill = Mean)) +
  geom_tile() + 
  scale_fill_viridis_c() +
  facet_grid(~synth_or_deg,scales = "free_x",space = "free_x") +
  labs(y = NULL, x= "Gene" , fill ="Mean Contig\nRead Coverage") +
  theme_bw() +
  scale_y_discrete(limits=rev) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())

ggsave("results/Ectoine_gene_coverage_to_Class_only_Halobacillus_ectGenes.png",width = 3, height = 1.5, dpi = 600, scale = 2.5)
ggsave("results/Ectoine_gene_coverage_to_Class_only_Halobacillus_ectGenes.svg",width = 3, height = 1.5, dpi = 600, scale = 2.5)


blast_res_filtered %>% 
  mutate(contig_taxonomy = str_remove(contig_taxonomy,"\\|c__.*"),
         contig_taxonomy = str_remove_all(contig_taxonomy,"\\|[a-z]__unassigned"),
         contig_taxonomy = str_replace_all(contig_taxonomy,"\\|o","\no"),
         contig_taxonomy = str_replace_all(contig_taxonomy,"\\|"," | "),
         contig_taxonomy = if_else(is.na(contig_taxonomy),"Unclassified",contig_taxonomy)) %>% 
  group_by(contig_taxonomy,geneID,synth_or_deg) %>% 
  summarise(RPKM = mean(RPKM),
            Mean = mean(Mean)) %>% ungroup() %>% group_by(contig_taxonomy) %>% 
  mutate(contig_taxonomy = if_else(contig_taxonomy == "k__Bacteria","Other Bacteria",contig_taxonomy),
         contig_taxonomy = str_remove(contig_taxonomy, "k__"),
         num_genes = sum(Mean > 0)) %>% 
  ggplot(aes(geneID, reorder(contig_taxonomy,num_genes) , fill = Mean)) +
  geom_tile() + 
  scale_fill_viridis_c() +
  facet_grid(~synth_or_deg,scales = "free_x",space = "free_x") +
  labs(y = NULL, x= "Gene" , fill ="Mean\nCoverage") +
  theme_bw()

ggsave("results/Ectoine_gene_coverage_to_Phylum.png",width = 3, height = 1.2, dpi = 600, scale = 3)

```

### Contig tax from mmseqs

(1) a single taxonomy numeric identifier
(2) a taxonomic rank column
(3) taxonomic name column, columns for the number of fragments: (3) retained, (4) taxonomically assigned, and (5) in agreement with the contig
label (i.e. same taxid or have it as an ancestor), (5) the support received.
```{r}
col_names <- c("ra_and_subs","reads_and_subs","reads","rank","NCBI_id","tax","full_lineage","lineage")

contig_tax_report <- read_tsv("data/contigs_LCA/S3_Fallen/assembly_LCA_report_w_standardized_lineage",col_names = col_names)

id_lineage_mapping <- contig_tax_report %>% select(NCBI_id, lineage)

col_names <- c("contig","NCBI_id","rank","tax","frags_retained","frags_tax_assigned","frags_agreeing_w_contig_label","support")

tax_levels <- c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")

contig_tax_mmseqs <- read_tsv("data/contigs_LCA/S3_Fallen/assembly_LCA_lca.tsv", col_names = col_names) %>% 
  left_join(id_lineage_mapping) %>% 
  separate(lineage, into = tax_levels, remove = FALSE,sep = ";") %>% 
  mutate(across(tax_levels, ~str_remove(.x, "^[a-z]__$")),
         lineage = if_else(tax == "unclassified", "unclassified", lineage),
         Kingdom = if_else(tax == "unclassified", "unclassified", Kingdom))


```


### Ectoine gene summary using contig taxonomy from UniRef LCA
```{r fig.height=4, fig.width=10}
gene_data <- read_csv("data/reference/ectoine_metabolism_genes.csv")

blast6_cols <- c("Query", "Subject", "Percent_ID", "Aln_Len", "Mismatch", "Gap", "Query_Start", "Query_Stop", "Subject_Start", "Subject_Stop", "E-Value", "Bit_Score", "percent_of_query_aligned", "subject_title","aligned_qeuery_seq", "aligned_subject_seq")

contig_coverage <- read_tsv("data/contig_coverage.tsv") %>% 
  rename_all(~str_remove(.x,".* "))

blast_res <- read_tsv("data/BLAST/ectoine_genes_BLAST.blastn", skip = 1,col_types = "ccdiddiiiidddccc", col_names = blast6_cols) %>% 
  left_join(gene_data) %>% 
  left_join(contig_tax_mmseqs %>% rename(subject_title = "contig")) %>% 
  left_join(contig_coverage %>% rename(subject_title = "Contig"))

# Filter using EggNog Mapper filtering criteria
blast_res_filtered <- blast_res %>% 
  filter(`E-Value` < 0.001, Bit_Score > 60, Percent_ID > 40, percent_of_query_aligned > 20) 

blast_res_filtered %>% 
  group_by(geneID,synth_or_deg, Kingdom, Phylum, Class, Order, Family) %>% 
  summarise(RPKM = mean(RPKM),
            Mean = mean(Mean),
            taxonomy = paste(Kingdom, Phylum, Class, Order, Family, sep = "; ") %>% str_remove_all("[; ]*$")) %>% 
  group_by(taxonomy) %>% filter(length(unique(geneID)) > 2, taxonomy != "k__Bacteria") %>% 
  ggplot(aes(geneID, str_wrap(taxonomy, width = 60), fill = Mean)) +
  geom_tile() + 
  scale_fill_viridis_c() +
  facet_grid(~synth_or_deg,scales = "free_x",space = "free_x") +
  labs(y = NULL, x= "Gene" , fill ="Mean\nCoverage") +
  theme_bw()

ggsave("results/Ectoine_gene_coverage_mmseqs.png",width = 4, height = 2, dpi = 600, scale = 3)
ggsave("results/Ectoine_gene_coverage_mmseqs.pdf",width = 4, height = 2, dpi = 600, scale = 3)
ggsave("results/Ectoine_gene_coverage_mmseqs.eps",width = 4, height = 2, dpi = 600, scale = 3)

```


## KOFAMscan
```{r}
cols <- c("sig","gene_name", "KO", "thrshld", "score","e_value","KO_definition")

kofams <- read_tsv("data/kofamScan_assembly/S3_Fallen/kofam_results.txt", col_names = cols, skip = 1) 

```

## UNIREF read mapping with mmseqs

```{r}
bakta_con <- DBI::dbConnect(RSQLite::SQLite(),"~/work/bakta/db/bakta.db")

bakta_db <- RSQLite::dbListTables(bakta_con)

unique_seqs <- tbl(bakta_con, "ups")

uniref100_info <- tbl(bakta_con, "ips")

uniref90_info <- tbl(bakta_con, "psc")
uniref50_info <- tbl(bakta_con, "pscc")

test <- unique_seqs %>% glimpse()
```



(1) Target identifier
(2) Number of sequences aligning to target
(3) Unique coverage of target uniqueAlignedResidues / targetLength
(4) Target coverage alignedResidues / targetLength
(5) Average sequence identity
(6) Taxonomical information identifier, species, lineage

### Get gene hits and annotations
```{r}
cols <- c("target", "num_seqs_aligned", "unique_coverage_of_target", "target_coverage", "average_seq_identity", "taxonomy")


uniref_read_mapping <- read_tsv("data/mmseqs/concreteMetaG_tophit_report",col_names = cols) %>% 
  mutate(uniref100_id = str_remove(target, "UniRef100_"))


uniref100_to_uniref90 <- uniref100_info %>% 
  filter(uniref100_id %in% local(uniref_read_mapping$uniref100_id)) %>% 
  left_join(uniref90_info,by = "uniref90_id") %>%
  left_join(uniref50_info,by = "uniref50_id") %>%
  collect()

uniref_read_maping_w_annotatinos <- uniref_read_mapping %>% 
  left_join(uniref100_to_uniref90) %>% 
  arrange(desc(target_coverage))

```


### Biofilm involved genes

```{r}
biofilm_kos_keggDecoder <- c("K11935","K11931", "K11936", "K11937", # pgaABCD; biofilm PGA synthesis protein
                             "K13654", # mcbR; GntR family transcriptional regulator, colanic acid and biofilm gene transcriptional regulator
                             "K12148", # bssS; biofilm regulator BssS
                             "K13650", #mcbA; MqsR-controlled colanic acid and biofilm protein A
                             "K04335","K04334", "K04336", #csgABC; curli fimbriae biosynthesis
                             "K12687" #flu; antigen 43
                             ) %>% 
  str_remove("^K")

biofilm_hits <- uniref_read_maping_w_annotatinos %>% 
  filter(if_any(everything(), ~str_detect(.x,"[b,B]iofilm")) | 
           if_any(everything(), ~.x %in% biofilm_kos_keggDecoder))

biofilm_hits_formatted <- biofilm_hits %>% 
  select(target, num_seqs_aligned, percent_of_gene_covered = unique_coverage_of_target, coverage_depth = target_coverage, taxa = X8, kegg_orthology_id, uniref90_product = product.y, uniref50_product = product) %>% 
  mutate(process = "Bioflims") %>% 
  relocate(process, taxa) %>% 
  write_tsv("results/biofilm_gene_hits.tsv")
```


### Anaerobic processes
```{r}

ferment_kos <- c("K00016", # pyruvate to lactate
                 "K00656", # pyruvate to formate + acetyle-CoA
                 "K00127", "K00124", "K00123", "K00126", "K00125", "K00122", # formate to co2 & h2
                 "K00156", # acetate from pyruvate,
                 "K00158","K01512", # acetate from pyruvate via acetyl-P
                 "K01067", # acetate from acetly-CoA
                 "K13788","K04020","K01512", #acetate from acetyl-CoA via acetly-P
                 "K00467", # acetate from lactate
                 "K00128", "K14085", "K00149", "K00129", "K00138", # ethanol step 1, acetaldehyde from acetate
                 "K04021", "K18366", "K04073", "K04072", "K00132" # ethanol step 1, acetabldehyde from acetyl-CoA
                 ) %>% str_remove("^K")

ferment_hits <- uniref_read_maping_w_annotatinos %>% 
  filter(if_any(everything(), ~.x %in% ferment_kos) | 
         if_any(everything(), ~str_detect(.x,"[A,a]lcohol[ ,-][d,D]ehydrogenase")))

ferment_hits_formatted <- ferment_hits %>% 
  select(target, num_seqs_aligned, percent_of_gene_covered = unique_coverage_of_target, coverage_depth = target_coverage, taxa = X8, kegg_orthology_id, uniref90_product = product.y, uniref50_product = product) %>% 
  mutate(process = "Fermentation")

srb_kos <- c("K11180", "K11181") %>% str_remove("^K")

srb_hits <- uniref_read_maping_w_annotatinos %>% 
  filter(if_any(everything(), ~.x %in% srb_kos))

dnra_kos <- c("K00363", "K00362", "K03385", "K15876") %>% str_remove("^K")

dnra_hits <- uniref_read_maping_w_annotatinos %>% 
  filter(if_any(everything(), ~.x %in% dnra_kos))

dnra_hits_formatted <- dnra_hits %>% 
  select(target, num_seqs_aligned, percent_of_gene_covered = unique_coverage_of_target, coverage_depth = target_coverage, taxa = X8, kegg_orthology_id, uniref90_product = product.y, uniref50_product = product)

nitrate_reductase_ec <- c("1.9.6.1")

nitrite_reductase_hits <- uniref_read_maping_w_annotatinos %>% 
  filter(if_any(everything(), ~.x %in% nitrate_reductase_ec))

nitrite_reductase_hits_formatted <- nitrite_reductase_hits %>% 
  select(target, num_seqs_aligned, percent_of_gene_covered = unique_coverage_of_target, coverage_depth = target_coverage, taxa = X8, kegg_orthology_id, uniref90_product = product.y, uniref50_product = product) %>% 
  mutate(process = "Nitrate reduction (dissimilatory)")

nitrite_reductase_kos <- c("K03385") %>% str_remove("^K")

nitrite_reductase <- uniref_read_maping_w_annotatinos %>% 
  filter(if_any(everything(), ~.x %in% nitrite_reductase_kos))

denit_kos <- c("K00376") %>% str_remove("^K")

denit_hits <- uniref_read_maping_w_annotatinos %>% 
  filter(if_any(everything(), ~.x %in% denit_kos))

denit_hits_formatted <- denit_hits %>% 
  select(target, num_seqs_aligned, percent_of_gene_covered = unique_coverage_of_target, coverage_depth = target_coverage, taxa = X8, kegg_orthology_id, uniref90_product = product.y, uniref50_product = product) %>% 
  mutate(process = "Denitrification")


formatted_table <- bind_rows(denit_hits_formatted,nitrite_reductase_hits_formatted,ferment_hits_formatted) %>% 
  relocate(process, taxa) %>% 
  write_tsv("results/anaerobic_processes_gene_hits.tsv")

```


## Cyanobacteria?
```{r}
cyano_hits <- uniref_read_maping_w_annotatinos %>% 
  filter(if_any(everything(), ~str_detect(.x,"[c,C]yanobacteria")))
```


