---
title: "Shogun stats"
output: html_notebook
---


```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir = here::here())
library(tidyverse)
library(metacoder)
```


Calculating what % of reads were classified by BURST alignment with SHOGUN.

## Shogun & Wol
```{r}
library(tidyverse)
 # 1.	 qseqid	 query (e.g., unknown gene) sequence id
 # 2.	 sseqid	 subject (e.g., reference genome) sequence id
 # 3.	 pident	 percentage of identical matches
 # 4.	 length	 alignment length (sequence overlap)
 # 5.	 mismatch	 number of mismatches
 # 6.	 gapopen	 number of gap openings
 # 7.	 qstart	 start of alignment in query
 # 8.	 qend	 end of alignment in query
 # 9.	 sstart	 start of alignment in subject
 # 10.	 send	 end of alignment in subject
 # 11.	 evalue	 expect value
 # 12.	 bitscore	 bit score

names <- c("qseqid", "sseqid", "pident", "length", "mismatch", "gapopen", "qstart", "qend", "sstart", "send", "evalue", "bitscore")

alignment <- read_tsv("data/shogun/burst/wol_burst_alignment.b6",col_names = names) %>% 
  select(qseqid) %>% mutate(aligner = "burst")

bt2_names <- c("qseqid", "flag", "rname", "strand", "pos", "qwidth", "mapq", "cigar", "mrnm", "mpos", "isize", "seq", "qual")
alignment_bowtie2 <- read_delim("data/shogun/bt2/alignment.bowtie2.sam", "\t", escape_double = FALSE, trim_ws = TRUE, col_names = bt2_names) %>% 
  select(qseqid) %>% mutate(aligner = "bt2")

align_merged <- bind_rows(alignment,alignment_bowtie2)

s3_reads <- align_merged %>% 
  filter(str_detect(qseqid,"S3_Fallen")) %>% 
  mutate(
    dir = case_when(
      str_detect(qseqid,"S3_Fallen_R1") ~ "F",
      str_detect(qseqid,"S3_Fallen_R2") ~ "R"
    ))

S3_f_total <- Biostrings::readDNAStringSet("data/qc_sequence_files/S3_Fallen_R1.fastq.gz",format = "fastq") %>% length()
S3_r_total <- Biostrings::readDNAStringSet("data/qc_sequence_files/S3_Fallen_R2.fastq.gz",format = "fastq") %>% length()

wol_s3_summary <- s3_reads %>% 
  group_by(dir, aligner) %>% 
  summarise(num_classified = n()) %>% 
  mutate(
    total_reads = case_when(
      dir == "F" ~ S3_f_total,
      dir == "R" ~ S3_r_total
    )) %>%
  mutate(percent_classified = num_classified/total_reads *100)

```

## Kraken metacodeR
```{r}
bracken_raw <- read_tsv("data/braken_full_tax.tsv",col_names = c("tax","count"))

kraken_raw <- read_tsv("data/kraken")

bracken <- bracken_raw %>%
  mutate(root = str_match(tax,"x_.*"),
         domain = str_match(tax,"d_.*"),
         phylum = str_match(tax,"p_.*"),
         class = str_match(tax,"c_.*"),
         order = str_match(tax,"o_.*"),
         family = str_match(tax,"f_.*"),
         genus = str_match(tax,"g_.*"),
         species = str_match(tax,"s_.*")) %>%
  mutate(across(.cols = any_of(c("root","domain","phylum","class","order","family", "genus", "species")), .fns =  ~str_remove(.x,pattern = "\\|.*")))
  
  
species_bracken <- bracken %>% 
  filter(!is.na(species))

```

```{r}
library(readr)
library(metacoder)
library(ggplot2)
library(dplyr)


obj <- parse_tax_data(species_bracken, class_cols = "tax", class_sep = "|",
                      class_key = c(tax_rank = "taxon_rank", tax_name = "taxon_name"),
                      class_regex = "^(.+)_(.*)$")

#converting to relative abundance
obj$data$tax_data <- calc_obs_props(obj, "tax_data")

#summing per-taxon counts
obj$data$tax_abund <- calc_taxon_abund(obj, "tax_data")

obj$data$tax_abund$S3 <- obj$data$tax_abund$count


#All domains
(tree <- obj %>%
    taxa::filter_taxa(S3 > 0.001) %>%
    filter_taxa(taxon_names %in% c("Viruses", "Bacteria", "Archaea", "Eukaryota"),
              subtaxa = TRUE) %>% 
    heat_tree(node_label = taxon_names,
          node_size = S3,
          node_size_range = c(0.002,0.04),
          node_label_size_range = c(.015,.025),
          node_size_axis_label = "OTU count",
          initial_layout = "reingold-tilford", layout = "davidson-harel",
          overlap_avoidance = 10,
          node_label_max = 60,
          node_color = S3,
          node_color_range = c("gray","yellow","red"),
          node_color_axis_label = "Relative abundance")
)
ggsave(plot = tree,filename = "results/figures/kraken2_metacoder.pdf", device = cairo_pdf, width = 12, height = 12, dpi = 600)


#Bacteria
(bacteria_tree <- obj %>%
    taxa::filter_taxa(S3 > 0.0001) %>%
    filter_taxa(taxon_names %in% c("Bacteria"),
              subtaxa = TRUE) %>% 
    heat_tree(node_label = taxon_names,
          node_size = S3,
          node_size_range = c(0.002,0.04),
          node_label_size_range = c(.015,.025),
          node_size_axis_label = "OTU count",
          initial_layout = "reingold-tilford", layout = "davidson-harel",
          overlap_avoidance = 10,
          node_label_max = 60,
          node_color = S3,
          node_color_range = c("gray","yellow","red"),
          node_color_axis_label = "Relative abundance")
)

ggsave(plot = bacteria_tree,filename = "results/figures/kraken2_bacteria_metacoder.pdf", device = cairo_pdf, width = 12, height = 12, dpi = 600)


#Archaea 
(archaea_tree <- obj %>%
    taxa::filter_taxa(S3 > 0.0001) %>%
    filter_taxa(taxon_names %in% c("Archaea"),
              subtaxa = TRUE) %>% 
    heat_tree(node_label = taxon_names,
          node_size = S3,
          node_size_range = c(0.002,0.04),
          node_label_size_range = c(.015,.025),
          node_size_axis_label = "OTU count",
          initial_layout = "reingold-tilford", layout = "davidson-harel",
          overlap_avoidance = 10,
          node_label_max = 60,
          node_color = S3,
          node_color_range = c("gray","yellow","red"),
          node_color_axis_label = "Relative abundance")
)
ggsave(plot = archaea_tree,filename = "results/figures/kraken2_archaea_metacoder.pdf", device = cairo_pdf, width = 12, height = 12, dpi = 600)


#Viruses
(virus_tree <- obj %>%
    filter_taxa(taxon_names %in% c("Viruses"),
              subtaxa = TRUE) %>% 
    taxa::filter_taxa(S3 > 0.00001) %>%
    heat_tree(node_label = taxon_names,
          node_size = S3,
          node_size_range = c(0.002,0.04),
          node_label_size_range = c(.015,.025),
          node_size_axis_label = "OTU count",
          initial_layout = "reingold-tilford", layout = "davidson-harel",
          overlap_avoidance = 10,
          node_label_max = 60,
          node_color = S3,
          node_color_range = c("gray","yellow","red"),
          node_color_axis_label = "Relative abundance")
)
ggsave(plot = virus_tree,filename = "results/figures/kraken2_virus_metacoder.pdf", device = cairo_pdf, width = 12, height = 12, dpi = 600)


#Eukaryotes
(euk_tree <- obj %>%
    taxa::filter_taxa(S3 > 0.00001) %>%
    filter_taxa(taxon_names %in% c("Eukaryota"),
              subtaxa = TRUE) %>% 
    heat_tree(node_label = taxon_names,
          node_size = S3,
          node_size_range = c(0.002,0.04),
          node_label_size_range = c(.015,.025),
          node_size_axis_label = "OTU count",
          initial_layout = "reingold-tilford", layout = "davidson-harel",
          overlap_avoidance = 10,
          node_label_max = 100,
          node_color = S3,
          node_color_range = c("gray","yellow","red"),
          node_color_axis_label = "Relative abundance")
)
ggsave(plot = euk_tree,filename = "results/figures/kraken2_euk_metacoder.pdf", device = cairo_pdf, width = 12, height = 12, dpi = 600)

```


## Updated Kraken using both GTDB & Refseq database
```{r}
library(tidyverse)


########
##Read in GTDB classified report
#######

cols <- c("ra_and_subs","reads_and_subs","reads","rank","NCBI_id","tax")

kraken <- read_tsv("data/kraken2/S3_Fallen_brackenReport.txt", 
                   col_names = cols,trim_ws = F)

ranks_and_num <- unique(kraken$rank) %>% 
  data.frame(rank = .) %>% 
  mutate(num_rank = row_number())

tax_split <- ranks_and_num %>% filter(rank != "U") %>% pull(rank) %>% as.character()

#Fill in higher level taxonomy and calculate relative abundances without sub-taxa
kraken_gtdb <- kraken %>% separate(tax,into = tax_split,sep = "  ") %>%
  mutate(across(where(is.character), ~na_if(.x,""))) %>%
  left_join(ranks_and_num) %>% 
  fill(any_of(tax_split)) %>% 
  mutate(taxa = row_number()) %>% 
  pivot_longer(tax_split,names_to = "row_rank",values_to = "row_tax") %>% 
  left_join(ranks_and_num %>% rename(row_rank = "rank",row_num_rank = "num_rank")) %>% 
  filter(row_num_rank <= num_rank) %>% 
  pivot_wider(id_cols = c("taxa","ra_and_subs","reads_and_subs","reads","rank","NCBI_id"), names_from = "row_rank",values_from = "row_tax") %>% 
  mutate(rel.abund = reads/sum(reads)) %>% 
  mutate(R = paste0("r__",R),
         taxonomy = paste(R,D,P,C,O,F,G,S,sep = "|"),
         taxonomy = str_remove(taxonomy,"[NA|]+$"))





########
##Read in refseq classified report
#######

kraken <- read_tsv("data/kraken2_euk/S3_Fallen_brackenReport.txt", col_names = cols,trim_ws = F)

tax_levels <- c("D","P","C","O","F","G","S")

refseq_kraken_from_std_tax_ranks <- read_tsv("data/kraken2_euk/S3_Fallen_brackenReport_w_standardized_lineage.txt", col_names = c(cols,"lineage","std_lineage"),trim_ws = F) %>% 
  separate(std_lineage, into = tax_levels, remove = FALSE,sep = ";") %>% 
  mutate(R = "r__root", 
         D = str_replace(D, "k__", "d__"),
         across(any_of(tax_levels), ~if_else(str_detect(.x, "^[a-z]__$"), NA, .x)),
         rel.abund = reads / sum(reads),
         taxonomy = paste(R,D,P,C,O,F,G,S,sep = "|"),
         taxonomy = str_remove(taxonomy,"[NA|]+$"))

#Combine bacteria and archaea from GTDB with other life from refseq

kraken_otu <- kraken_gtdb %>% 
  select(reads, rel.abund, taxonomy) %>% 
  bind_rows(refseq_kraken_from_std_tax_ranks %>% filter(!D %in% c("d__Archaea", "d__Bacteria")) %>% filter(!if_any(everything(), ~str_detect(.x,"\\|NA\\|"))) %>% 
              select(reads, rel.abund, taxonomy)) %>% 
  mutate(rel.abund = reads/sum(reads)) %>% 
  filter(rel.abund > 0)


kraken_taxa_split <- kraken_otu %>% separate(taxonomy,into = c("root","kingdom","phylum","class","order","family","genus","species"),sep = "\\|")

kingdom <- kraken_taxa_split %>% group_by(kingdom) %>% summarise(rel.abund = sum(rel.abund))
  
phylum <- kraken_taxa_split %>% group_by(kingdom,phylum) %>% summarise(rel.abund = sum(rel.abund))
  
class <- kraken_taxa_split %>% group_by(kingdom, phylum, class) %>% summarise(rel.abund = sum(rel.abund))
  
order <- kraken_taxa_split %>% group_by(kingdom, phylum, class, order) %>% summarise(rel.abund = sum(rel.abund))
  
family <- kraken_taxa_split %>% group_by(kingdom, phylum, class, order, family) %>% summarise(rel.abund = sum(rel.abund))
  
genus <- kraken_taxa_split %>% group_by(kingdom, phylum, class, order, family, genus) %>% summarise(rel.abund = sum(rel.abund))
  
species <- kraken_taxa_split %>% group_by(kingdom, phylum, class, order, family, genus) %>% summarise(rel.abund = sum(rel.abund))


kraken_otu_genus <- genus %>% 
  mutate(taxonomy = paste("r__root",kingdom, phylum, class, order, family, genus,sep = "|"),
         taxonomy = str_remove(taxonomy,"[NA|]+$"))

```


### Plot the merged table with metacodeR
```{r}
obj <- parse_tax_data(kraken_otu_genus, class_cols = "taxonomy", class_sep = "|",
                      class_key = c(tax_rank = "taxon_rank", tax_name = "taxon_name"),
                      class_regex = "^(.+)__(.*)$")

#converting to relative abundance
obj$data$tax_data <- calc_obs_props(obj, "tax_data")

#summing per-taxon counts
obj$data$tax_abund <- calc_taxon_abund(obj, "tax_data")

obj$data$tax_abund$S3 <- obj$data$tax_abund$rel.abund


(tree <- obj %>%
    filter_taxa(S3 > 0.001,supertaxa = TRUE, reassign_obs = TRUE) %>%
    heat_tree(node_label = taxon_names,
          node_size = S3,
          node_size_range = c(0.00175,0.045),
          node_label_size_range = c(.015,.025),
          node_size_axis_label = "OTU count",
          initial_layout = "reingold-tilford", layout = "davidson-harel",
          overlap_avoidance = 10,
          node_label_max = 75,
          node_color = S3,
          node_color_range = c("gray","gray","gray"),
          node_color_axis_label = "Relative abundance")
)
ggsave(plot = tree,filename = "results/figures/finerkraken2_gtdb_combined.pdf", device = cairo_pdf(), width = 12, height = 12, dpi = 600)


```


## UniRef LCA taxonomy report from MMseqs2


```{r}
cols <- c("ra_and_subs","reads_and_subs","reads","rank","NCBI_id","tax","full_lineage","lineage")

mmseqs <- read_tsv("data/mmseqs/concreteMetaG_report_w_standardized_lineage", 
                   col_names = cols,trim_ws = F)

mmseqs_for_metacoder <- mmseqs %>% 
  mutate(rel_abund = reads / sum(reads)) %>% 
  filter(!tax %in% c("unclassified","  cellular organisms")) %>% 
  mutate(lineage = str_glue("r__Root;{lineage}"),
         lineage = str_remove_all(lineage, "(;[a-z]__)*$")) %>% 
  filter(!str_detect(lineage, ";NA")) %>% 
  select(rel_abund, lineage)

obj <- parse_tax_data(mmseqs_for_metacoder, class_cols = "lineage", class_sep = ";",
                      class_key = c(tax_rank = "taxon_rank", tax_name = "taxon_name"),
                      class_regex = "^(.+)__(.*)$")

#summing per-taxon counts
obj$data$tax_abund <- calc_taxon_abund(obj, "tax_data")

obj$data$tax_abund$S3 <- obj$data$tax_abund$rel_abund


#All domains
(tree <- obj %>%
    filter_taxa(obj$data$tax_abund$rel_abund > 0.001,supertaxa = TRUE, reassign_obs = TRUE) %>%
    heat_tree(node_label = taxon_names,
          node_size = S3,
          node_size_range = c(0.00175,0.045),
          node_label_size_range = c(.015,.025),
          node_size_axis_label = "OTU count",
          initial_layout = "reingold-tilford", layout = "davidson-harel",
          overlap_avoidance = 10,
          node_label_max = 75,
          node_color = S3,
          node_color_range = c("gray","gray","gray"),
          node_color_axis_label = "Relative abundance")
)
ggsave(plot = tree,filename = "results/figures/mmseqs2_metacoder.pdf", device = cairo_pdf, width = 12, height = 12, dpi = 600)



```




```{r}
########
##Read in GTDB classified report
#######

cols <- c("ra_and_subs","reads_and_subs","reads","rank","NCBI_id","tax")

mmseqs <- read_tsv("data/mmseqs/concreteMetaG_report", 
                   col_names = cols,trim_ws = F)

report <- pavian::read_report("data/mmseqs/concreteMetaG_report")

ranks <- report$taxRank %>% unique()

ranksto_use <- c("D","K","P","C","O","F","G","S")

filtered_report <- report %>% 
  filter(taxRank %in% ranksto_use) %>% 
  pivot_wider(names_from = taxRank, values_from = name) %>% 
  fill(D,.direction = "down") %>%
  group_by(D,.drop = FALSE) %>% 
  fill(K,.direction = "down") %>%
  group_by(K,.drop = FALSE) %>% 
  fill(P,.direction = "down") %>%
  group_by(P,.drop = FALSE) %>% 
  fill(C,.direction = "down") %>%
  group_by(C,.drop = FALSE) %>% 
  fill(O,.direction = "down") %>%
  group_by(O,.drop = FALSE) %>% 
  fill(F,.direction = "down") %>%
  group_by(F,.drop = FALSE) %>% 
  fill(G,.direction = "down") %>%
  group_by(G,.drop = FALSE) %>% 
  fill(S,.direction = "down")

tax_split <- ranks_and_num %>% filter(rank != "U") %>% pull(rank) %>% as.character()

#Fill in higher level taxonomy and calculate relative abundances without sub-taxa
mmseqs_parsed <- mmseqs %>% separate(tax,into = tax_split,sep = "  ") %>%
  mutate(across(everything(), ~na_if(.x,""))) %>%
  left_join(ranks_and_num) %>% 
  fill(any_of(tax_split)) %>% 
  mutate(taxa = row_number()) %>% 
  pivot_longer(tax_split,names_to = "row_rank",values_to = "row_tax") %>% 
  left_join(ranks_and_num %>% rename(row_rank = "rank",row_num_rank = "num_rank")) %>% 
  filter(row_num_rank <= num_rank) %>% 
  pivot_wider(id_cols = c("taxa","ra_and_subs","reads_and_subs","reads","rank","NCBI_id"), names_from = "row_rank",values_from = "row_tax") %>% 
  mutate(rel.abund = reads/sum(reads)) #%>% 
  mutate(R = paste0("r__",R),
         taxonomy = paste(R,D,P,C,O,F,G,S,sep = "|"),
         taxonomy = str_remove(taxonomy,"[NA|]+$"))

```






